allennlp/semparse/worlds/atis_world.py

```diff
@@ -27,34 +51,72 @@ def __init__(self, utterances: List[str], tokenizer=None) -> None:
+    def init_all_valid_actions(self) -> Tuple[Dict[str, List[str]], numpy.ndarray]:
         """
-        We initialize the world's valid actions with that of the context. This means that the strings
-        and numbers that were valid earlier in the interaction are also valid. We then add new valid strings
-        and numbers from the current utterance.
+        We initialize the valid actions with the global actions. We then iterate through the
+        utterances up to and including the current utterance and add the valid strings.
         """
+
         valid_actions = deepcopy(self.sql_table_context.valid_actions)
-        for string in self.get_strings_from_utterance():
+        linking_scores = []
+        current_tokenized_utterance = [] if not self.tokenized_utterances \
+                else self.tokenized_utterances[-1]
+
+        strings: Set[str] = set()
+        for tokenized_utterance in self.tokenized_utterances:
+            string_linking_dict = get_strings_from_utterance(tokenized_utterance)
+            strings = strings.union(set(string_linking_dict.keys()))
+
+        strings_list: List[str] = sorted(strings, reverse=True)
+
+        # We construct the linking scores for strings from the ``string_linking_dict`` here.
+        string_linking_scores = []
+        for string in strings_list:
+            entity_linking = [0 for i in range(len(current_tokenized_utterance))]
+            if string in string_linking_dict:
+                for idx in string_linking_dict[string]:
+                    entity_linking[idx] = 1
+            string_linking_scores.append(entity_linking)
+        linking_scores.extend(string_linking_scores)
+
+        for string in strings_list:
             action = format_action('string', string)
             if action not in valid_actions['string']:
                 valid_actions['string'].append(action)
 
-        numbers = ['0', '1']
-        for utterance in self.utterances:
-            numbers.extend(get_numbers_from_utterance(utterance))
-            for number in numbers:
-                action = format_action('number', number)
-                if action not in valid_actions['number']:
-                    valid_actions['number'].append(action)
-
-        return valid_actions
+        numbers = {'0', '1'}
+        number_linking_dict: Dict[str, List[int]] = {}
+
+        for idx, (utterance, tokenized_utterance) in enumerate(zip(self.utterances, self.tokenized_utterances)):
+            number_linking_dict = get_numbers_from_utterance(utterance, tokenized_utterance)
+            numbers = numbers.union(set(number_linking_dict.keys()))
+        numbers_list: List[str] = sorted(numbers, reverse=True)
+
+        # We construct the linking scores for numbers from the ``number_linking_dict`` here.
+        number_linking_scores = []
+        for number in sorted(numbers_list, reverse=True):
+            entity_linking = [0 for i in range(len(current_tokenized_utterance))]
```

Comment:
Similarly here: `entity_linking = [0 for token in current_tokenized_utterance]`, and the other comments on the for loop.  Makes me wonder if the shared logic could be pulled into a method.  Not sure it's worth it, as the treatment of the two might change.